### ISSUE TITLE
Proposal: default struct fields to zeros when translated from C

### ISSUE THREAD

**Issue by McSinyx** [ziglang/zig#8165]
[Accepted: `translate-c` produces structs with zero as default field values.](https://github.com/ziglang/zig/issues/8165#issuecomment-1229098623)

---

In C structs, omitted elements are initialized to zeros, and thus some C libraries declare structs to be used this way.  However, Zig struct literals do not allow field omission, and thus [initializing structs defined in C can be quite verbose](https://git.sr.ht/~cnx/zent/commit/dff21276b87f03f6efe8f497976887074b8e4674).

Per the documentation, `extern struct`, which the C struct got translated to, [should only be used for compatibility with the C ABI](https://ziglang.org/documentation/master/#extern-struct).  On the other hand, `extern struct` when used in C still have the defaulting-to-zeros behavior, so I'm suggesting to mirror this in Zig as well.  In other words, the following test should pass instead of failing to compile.

```zig
const expect = @import("std").testing.expect;

const Point = extern struct {
    x: c_int,
    y: c_int,
};

test "extern struct init" {
    const p = Point{ .x = 42 };
    expect(p.x == 42);
    expect(p.y == 0);
}
```

In case of nested structs, the initialization should be performed recursively as well.

**Comment by Tetralux** [comment]
FYI, the way to accomplish this currently, is to give all optional fields a default value of `0`, which allows you to omit those fields; or use `mem.zeroes(Point)` and set the fields you want.

**Comment by tadeokondrak** [comment]
I think this would be a good improvement to translate-c rather than a language change.

**Comment by McSinyx** [comment]
> FYI, the way to accomplish this currently, is to give all optional fields a default value, which allows you to omit those fields

I'm well aware of that, but it's not feasible in the case of `cInclude`d headers, which are translated to Zig under the hood.

> I think this would be a good improvement to translate-c rather than a language change.

Yes, and maybe no.  IMHO the calling interface for a C ABI struct should be the same, regardless if it is defined in C or in Zig as `extern struct`.  To elaborate, if `struct Point { int x, y; }` can be initialized via `struct Point p = { .x = 42 }` in C99 and `const p = Point{ .x = 42 }` in (proposed) Zig, then `const Point = extern struct { x: c_int, y: c_int }` which can also be used in the same manner in C99 should be initializable via `const p = Point{ .x = 42 }` in (proposed) Zig.

Of course, Zig can translate the C struct to `const Point = extern struct { x: c_int = 0, y: c_int = 0 }` and suggest the convention for the latter case, but I feel that manually constructing such struct is redundant.  Furthermore, regarding backward compatibility, AFAICT implicitly defaulting `extern struct` fields to zeros does not break anything.

**Comment by InterplanetaryEngineer** [comment]
> I think this would be a good improvement to translate-c rather than a language change.

I second this and would see *implicit* zero-initialization as hidden behaviour, which is a major anti-goal of the core lagnuage.

**Comment by matu3ba** [comment]
## 1. Change of code semantics, when translating c files individually

Think of a user translating __file1.c__
```C
struct _s1
{
  int t1
} s1;
//
s1 defaultinit;
defaultinit.t1 = 0;
```

Now we have another file __file2.c__ which __does not use default initialisation__
```C
//
s1 defaultinit;
```

To prevent this, translate-c must view all includes of the struct and what they do => makes translation-c slower (to be always correct) or creates needlessly wrong zig code.

2. The most reliable translation approach is to __only translate the headers__ and generating wrong semantics is especially bad, when "it does not work all the time".

**Comment by jayschwa** [comment]
I am exploring the feasibility of converting an existing C codebase over to Zig. Some of the APIs lean heavily on C's implicit initialization to zero. If `@cImport` gave struct fields a default of zero, it would be trivial to start hand-converting initializers from C to Zig. As-is, it doesn't feel worth it. I hope this proposal (as it relates to C translation, not `extern struct`) is reconsidered.

**Comment by andrewrk** [comment]
I would be in favor of having translate-c produce structs with zero as default field values.

**Comment by jayschwa** [comment]
I'm glad to hear it. This sounds like a good opportunity for me to start learning my way around `translate_c.zig` 🕵🏻‍♂️ .

**Comment by andrewrk** [comment]
I think this can sit on "accepted" for now :+1:  but to be clear I do want to give @Vexu and @ehaas a chance to veto.

**Comment by ehaas** [comment]
I don't think doing this has any correctness problems, just trades (potentially) a bit of performance for a nicer API.

The tricky bit will be unions - empty-initialized unions perform empty-initialization on the first member. So the following code:
```c
struct S {
    int x;
    union {
        int i;
        float f;
    } y;
};
```
Will have to translate to:
```zig
const union_unnamed_1 = extern union {
    i: c_int,
    f: f32,
};
pub const struct_S = extern struct {
    x: c_int = 0,
    y: union_unnamed_1 = .{
        .i = 0,
    },
};
```

In C, empty initialization of a struct should also set padding bits to zero. Do we have any way to control that?

**Comment by Vexu** [comment]
I think a better and more explicit solution would be adding a `zero` member to all structs and unions:
```zig
pub const zero = @import("std").mem.zeroes(@This())
```
This would avoid the performance issues and also set the padding to zero.

**Comment by jayschwa** [comment]
Here is a pseudo-example of a pattern where default field values makes manual translation easy:

``` c
// original.c

#include <struct_def.h>

const BigStruct dudes[] = {
    {.foo = 1},
    {.bar = 2},
    {.baz = 3},
};
```

``` zig
// translated.zig (by human) from original.c

const c = @cImport({
    @cInclude("struct_def.h");
});

const dudes = [_]c.BigStruct{
    .{.foo = 1},
    .{.bar = 2},
    .{.baz = 3},
};
```

What would `translated.zig` look like with the alternative `zero` member?

What are the risks if padding is not zeroed out? It could interfere with (dubious) use of `memcmp`. What else?

**Comment by Vexu** [comment]
For something like that you should use `std.mem.zeroInit`:

**Comment by jayschwa** [comment]
I recently finished converting an ~18k SLOC project from C to Zig. I am sharing my experience (as it relates to this proposal) while it is still fresh in my head:

The conversion process was roughly:

1. Setup a `build.zig` that uses all of the existing C source files.
2. Rewrite one C source file at a time into Zig. This was mostly top-down (e.g. I started from `main`).
3. Once a C header file was no longer included by any C source file (because they had been converted to Zig), the C types from the header were then ported into Zig.

During step 2, the majority of structures I used were being included through `@cImport`. Use of `zeroInit` was pretty much mandatory during this transition period. It was perfectly functional, but kind of clunky. Once the structures were converted to native Zig types in step 3, I could remove the `zeroInit` calls.

In a couple instances, I forgot to remove a `zeroInit` call after converting to a native Zig structure. When I added a new field to the structure, `zeroInit` quietly initialized the new field to an invalid state. Without the `zeroInit`, the compiler would have helpfully complained to me about the uninitialized field.

In summary, I was able to complete the C to Zig conversion without this proposal implemented. However, I believe the process would have been smoother if I did not have to add, and then remove, calls to `zeroInit` along the way.

**Comment by InKryption** [comment]
> In a couple instances, I forgot to remove a zeroInit call after converting to a native Zig structure. When I added a new field to the structure, zeroInit quietly initialized the new field to an invalid state. Without the zeroInit, the compiler would have helpfully complained to me about the uninitialized field.

> In summary, I was able to complete the C to Zig conversion without this proposal implemented. However, I believe the process would have been smoother if I did not have to add, and then remove, calls to zeroInit along the way.

These two statements appear to be in contradiction with each other. If translate-c default-initialized fields to zero, would you not have run into the same problem as with `zeroInit` doing so, with it silently initializing newly-added fields to an invalid state (zero)?

**Comment by jayschwa** [comment]
> If translate-c default-initialized fields to zero, would you not have run into the same problem as with `zeroInit` doing so, with it silently initializing newly-added fields to an invalid state (zero)?

No, because the point at which I tripped over a lingering `zeroInit` was after I had ported the structure into Zig and was no longer using `@cImport`. I'll try to illustrate with a small pseudo-example:

---

Step 1 (all C source and headers)

``` c
// monster.h

typedef struct {
    Animation *attack;
    Animation *dance;
    Animation *pain;
    Animation *die;
} Monster;
```
``` c
// monster.c

void foo(void) {
    Monster yeti = {
        .dance = dance_anim,
        .die = die_anim,
    };
}
```

---

Step 2 (Zig source and C headers)

``` c
// monster.h

typedef struct {
    Animation *attack;
    Animation *dance;
    Animation *pain;
    Animation *die;
} Monster;
```
``` zig
// monster.zig

const c = @cImport({
    @cInclude("monster.h");
});

fn foo() void {
    const yeti = zeroInit(c.Monster, .{
        .dance = dance_anim,
        .die = die_anim,
    });
}
```

---

Ideal Step 3 (Pure Zig)

``` zig
// monster.zig

pub const Monster = struct {
    attack: ?*Animation = null,
    dance: ?*Animation = null,
    pain: ?*Animation = null,
    die: ?*Animation = null,
};

fn foo() void {
    const yeti: Monster = .{
        .dance = dance_anim,
        .die = die_anim,
    };
}
```

---

What I accidentally did for Step 3 in a couple spots:

``` zig
// monster.zig

pub const Monster = struct {
    attack: ?*Animation = null,
    dance: ?*Animation = null,
    pain: ?*Animation = null,
    die: ?*Animation = null,
};

fn foo() void {
    // Still using `zeroInit`, but there's no harm for now!
    const yeti = zeroInit(Monster, .{
        .dance = dance_anim,
        .die = die_anim,
    });
}
```

---

When the accidental, lingering `zeroInit` caused a problem:

``` zig
// monster.zig

pub const Monster = struct {
    fancy: FancyType, // Use `FancyType.init`

    attack: ?*Animation = null,
    dance: ?*Animation = null,
    pain: ?*Animation = null,
    die: ?*Animation = null,
};

fn foo() void {
    // Oops, `zeroInit` is quietly initializing `fancy`! 💥
    const yeti = zeroInit(Monster, .{
        .dance = dance_anim,
        .die = die_anim,
    });
}
```

---

With this proposal implemented, step 2 could have looked like this instead:

``` zig
// monster.zig

const c = @cImport({
    @cInclude("monster.h");
});

fn foo() void {
    const yeti: c.Monster = .{
        .dance = dance_anim,
        .die = die_anim,
    };
}
```

Later on, after adding `fancy: FancyType` to `Monster`, the compiler would have been like, "Hey, dummy, you need to initialize this field!"

Ultimately, me forgetting to remove the `zeroInit` was my bad, but I also wouldn't have needed to add the `zeroInit` in step 2 with this proposal.
